pyquil/api/_abstract_compiler.py

```diff
@@ -13,28 +13,92 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 ##############################################################################
+import sys
 from abc import ABC, abstractmethod
 from typing import Any, Dict, List, Optional, Sequence, Union
 
-from rpcq.messages import QuiltBinaryExecutableResponse, PyQuilExecutableResponse
-
+from rpcq.messages import (
+    NativeQuilRequest,
+    TargetDevice,
+    ParameterSpec,
+    ParameterAref,
+)
+
+from pyquil.api import Client
+from pyquil.api._error_reporting import _record_call
+from pyquil.device import AbstractDevice
+from pyquil.parser import parse_program
 from pyquil.paulis import PauliTerm
 from pyquil.quil import Program
+from pyquil.quilatom import MemoryReference, ExpressionDesignator
 from pyquil.quilbase import Gate
+from pyquil.version import __version__
+
+if sys.version_info < (3, 7):
+    from rpcq.external.dataclasses import dataclass
+else:
+    from dataclasses import dataclass
+
+
+class QuilcVersionMismatch(Exception):
+    pass
+
+
+class QuilcNotRunning(Exception):
+    pass
+
+
+@dataclass()
+class EncryptedProgram:
+    """
+    Encrypted binary, executable on a QPU.
+    """
+
+    program: str
+    """String representation of an encrypted Quil program."""
+
+    memory_descriptors: Dict[str, ParameterSpec]
+    """Descriptors for memory executable's regions, mapped by name."""
+
+    ro_sources: Dict[MemoryReference, str]
+    """Readout sources, mapped by memory reference."""
+
+    recalculation_table: Dict[ParameterAref, ExpressionDesignator]
+    """A mapping from memory references to the original gate arithmetic."""
+
+
+QuantumExecutable = Union[EncryptedProgram, Program]
```

Comment:
This seems semantically muddy to me. 

I'd consider something like renaming `Program` to `UnencryptedProgram` or `RawProgram` and `QuantumExecutable` to `Program`. Or leave `Program` as is and call this type `MaybeEncryptedProgram`.

Otherwise, I wouldn't object to using the verbose `Union[EncryptedProgram, Program]` directly as needed.

Reply:
The key thing being conveyed is that this things is "executable" -- that is, the raw program has passed through some form of compilation. It so happens that in the QVM scenario, compilation is optional, resulting in `Program` for that case.

I'm in favor of keeping this, as it was nice to have a single place to go when the definition of what is executable changes. This used to be (I think it was over near `QuantumComputer`):

```
ExecutableDesignator = Union[QuiltBinaryExecutableResponse, PyQuilExecutableResponse]
```